### ISSUE TITLE
duplicate negative TARGETID for stuck positioners

### ISSUE THREAD

**Issue by sbailey** [desihub/fiberassign#385]
## Duplicate negative TARGETIDs = bug

@AlexGKim discovered that we have duplicate negative TARGETIDs at different locations on different tiles, when TARGETIDs are supposed to be unique (even the negative ones).  This is due to a bug in [fiberassign.assign.write_assignment_fits_tile lines 308 - 313]( https://github.com/desihub/fiberassign/blob/802de4c853c29d104a676d925b74bda177989ab6/py/fiberassign/assign.py#L308-L313) where it is assigning negative TARGETIDs to unassigned (usually stuck) positioners that aren't assigned to a pre-identified target with a unique positive TARGETID.

The current logic boils down to
```
TARGETID = -(TILEID * 5020 + LOCATION)
```

The problem is that `LOCATION = 1000*PETAL_LOC + DEVICE_LOC` and thus has a max value of 9526 not 5020.  i.e. between tiles we should be stepping by `9526=max(locs)` instead of `5020=len(locs)`.

This would be fixable by changing

[assign.py line 310](https://github.com/desihub/fiberassign/blob/802de4c853c29d104a676d925b74bda177989ab6/py/fiberassign/assign.py#L310) from
```
unassign_offset = tile_id * nloc
```
to
```
unassign_offset = tile_id * 10000
```
or to avoid a hardcode 10000 this could be
```
unassign_offset = tile_id * np.max(locs)
```
though that has a different fragility of assuming that the set of possible locations won't ever change (probably valid, but could break if e.g. we permanently lost petal 9 and started fiberassigning only with petals 0-8 in the hardware model).

## Existing files

Also: we have a bunch of observed fiberassign files on disk with duplicate negative TARGETIDs on adjacent TILEID numbers.  e.g. the original case that Alex found:
```
 TARGETID  TILEID LOCATION FIBER
  int64    int64   int32   int32
---------- ------ -------- -----
-104206559  20757     6419  3061
-104206559  20758     1399   505
```
but there are many many more such cases.  I think we need to patch these in svn with different and genuinely unique negative TARGETIDs, then after desihub/desispec#1296 is implemented to allow svn to override the raw data fiberassign files (needed for other reasons), these will be picked up correctly for Fuji (i.e. this becomes semi-urgent).  Stating the obvious: any surgery of already-observed fiberassign files is dangerous, so proceed with caution.

## Related

Note that `desitarget.targets.encode_negative_targetid` is a completely different way of encoding an RA,dec into a unique negative TARGETID, but I contributed that too late and it was never used by fiberassign.  We could switch to using that instead.  That code is designed to never assign a negative TARGETID that would have been created by the current fiberassign logic.

Mentioning @schlafly @djschlegel @araichoor @dstndstn @tskisner to get this on the surveyops radar to fix prior to restarting in a couple weeks.

**Comment by schlafly** [comment]
If we can get this into fiberassign for a new tag on Monday that would be great.  If I should take this, I can figure it out as well.

**Comment by sbailey** [comment]
Yes, please go ahead and implement this and include it for the next tag.  I think the only thing needed is to change line 310:
```
unassign_offset = tile_id * nloc
```
to
```
unassign_offset = tile_id * 10000
```

I'm mostly offline for the weekend, so please don't wait around for a formal review from me.  Mentioning @dstndstn who I think was going to take care of patching the existing files with the same update.

**Comment by araichoor** [comment]
for info, there already is a draft version of the patched files ([desi-survey 2962]).

**Comment by schlafly** [comment]
Okay, I'll take this.  I said Monday, but I meant Tuesday; I'll also be mostly gone for the weekend.

**Comment by schlafly** [comment]
I'll merge this
https://github.com/desihub/fiberassign/pull/397
early tomorrow.  It's just Stephen's one liner from above.  Stating the obvious, this will break backward compatibility in that new and old fiberassign files will be different.  But we're patching the old fiberassign files and this doesn't affect any movement of positioners on the mountain, so we're okay with that.

**Comment by sbailey** [comment]
Fixed by PR #397.  Closing.
